artemis-commons/src/main/java/org/apache/activemq/artemis/utils/actors/BoundActor.java

```diff
@@ -0,0 +1,79 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.activemq.artemis.utils.actors;
+
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.function.ToIntFunction;
+
+public class BoundActor<T> extends ProcessorBase<Object> {
+
+   private static final AtomicIntegerFieldUpdater<BoundActor> SIZE_UPDATER = AtomicIntegerFieldUpdater.newUpdater(BoundActor.class, "size");
+   private volatile int size = 0;
+
+   private static final AtomicIntegerFieldUpdater<BoundActor> SCHEDULED_FUSH_UPDATER = AtomicIntegerFieldUpdater.newUpdater(BoundActor.class, "scheduledFlush");
+   private volatile int scheduledFlush = 0;
+
+   private static final Object FLUSH = new Object();
+
+
+   final int maxSize;
+   final ToIntFunction<T> sizeGetter;
+   ActorListener<T> listener;
+
+   Runnable overLimit;
+   Runnable clearLimit;
+
+   public BoundActor(Executor parent, ActorListener<T> listener, int maxSize, ToIntFunction<T> sizeGetter, Runnable overLimit, Runnable clearLimit) {
+      super(parent);
+      this.listener = listener;
+      this.maxSize = maxSize;
+      this.sizeGetter = sizeGetter;
+      this.overLimit = overLimit;
+      this.clearLimit = clearLimit;
+   }
+
+   @Override
+   protected final void doTask(Object task) {
+      if (task == FLUSH) {
+         this.scheduledFlush = 0;
```

Comment:
`scheduledFlush` must be `0` after `clearLimit::run` or a racy `act` after this would call `overLimit::run` before it

Reply:
@clebertsuconic this must be change to save bad things to happen concurrently

Reply:
it's a single thread setting this to false.. there's no concurrency here.

Reply:
`act` can set compareAndSet(0, 1) right after this `this.scheduledFlush = 0` meaning that the Runnable will run without the right order

Reply:
Right... the flush should set it to 0 no matter what.

Reply:
```
tasks = {1, 2, 3, F}
size = 3
max_size = 2
schedule_flush = 1

act(4)				|	actor's doTask) 	
--------------------------------|-------------------------------
size: 3 -> 4			|				
				|	tasks = {1, 2, 3, F} - {F}
				|	schedule_flush: 0
				|	
				|				
schedule_flush: 0 -> 1		|				
				|				
overLimit.run()			|				
				|	clearThreshold.run();
tasks = {} + {F}		|	
tasks = {} + {4}		|	
				|	tasks = {1, 2, 3} - {3}
				|	size: 4 -> 3
				```
				
				this is to help the future me on this :D

Reply:
that's actually why I set it to 0 before calling...

I may eventually have two in flight if I get one packet arriving with more size then the whole actor...


If you think there's a bug in there feel free to send a new PR *after* this one is merged.. I need to merge this asap.

Reply:
No bud this is a bug, because that listeners calls are not supposed to run in random order, but in the right sequence

Reply:
@franz1981 Ok, I am merging this now, If you want to improve this.. or talk to me next week.

Reply:
Ok, but I don't understand why you don't consider it a bug: listeners must not be called concurrently right? The compare and set is there to prevent it

Reply:
I added a test for this condition.. I don't think I would have caused any harm on the openwire flow control... I'm just trying to fix the flow control in openwire now.

Reply:
I know but given that is a racy behaviour it won't happen easily, like any concurrent bugs, but the execution I shown is valid and can still happen so worth to fix it.
At worst it would still fix your issue: run your existing tests and if it pass it means that is still working as you expect, if not, feel free to merge it as is now and I'll find out later what can be wrong (if any).

Reply:
I believe its fixed.